addons/mail/wizard/mail_compose_message.py

```diff
@@ -388,23 +389,30 @@ def onchange_template_id(self, template_id, composition_mode, model, res_id):
                 signature = self.env.user.signature
                 values['body_html'] = tools.append_content_to_html(values['body_html'], signature, plaintext=False)
         elif template_id:
+            keep = self.env['mail.template'].browse(template_id).report_template.attachment
             values = self.generate_email_for_composer(template_id, [res_id])[res_id]
             # transform attachments into attachment_ids; not attached to the document because this will
             # be done further in the posting process, allowing to clean database if email not send
-            attachment_ids = []
             Attachment = self.env['ir.attachment']
             for attach_fname, attach_datas in values.pop('attachments', []):
-                data_attach = {
-                    'name': attach_fname,
-                    'datas': attach_datas,
-                    'datas_fname': attach_fname,
-                    'res_model': 'mail.compose.message',
-                    'res_id': 0,
-                    'type': 'binary',  # override default_type from context, possibly meant for another model!
-                }
-                attachment_ids.append(Attachment.create(data_attach).id)
-            if values.get('attachment_ids', []) or attachment_ids:
-                values['attachment_ids'] = [(5,)] + values.get('attachment_ids', []) + attachment_ids
+                if keep:
+                    att = Attachment.search([
+                        ('name','=', attach_fname),
+                        ('res_model','=', model),
+                        ('res_id', '=', res_id)
+                    ])
+                if not keep or not att:
+                    data_attach = {
+                        'name': attach_fname,
+                        'datas': attach_datas,
+                        'datas_fname': attach_fname,
+                        'res_model': 'mail.compose.message',
+                        'res_id': 0,
+                        'type': 'binary',  # override default_type from context, possibly meant for another model!
+                    }
+                    att = Attachment.create(data_attach)
+                if att:
+                    attachment_ids.append(att.id)
```

Comment:
This assumes the previous `search` returned at most one result. Once again, not sure this is reliable.